### DISCUSSION TITLE
How to get the ID of a StateMachine object

### DISCUSSION THREAD

**Discussion by GuillaumeLeclerc** [micropython/micropython#11595] [Ideas]
Hello,

How can I get the id of the state machine that triggered a PIO IRQ. It seems that the only argument passed is the state machine that triggered it. But it's kind of useless because we have no attribute available to tell us which one it actually is.

Maybe it would be good to have an attribute. It seems that the data is available from the backend since `str(machine)` is working.

**Comment by jimmo** [comment]
Yes this would probably be useful, and in general on most types in the `machine` module -- e.g. going from a Pin instance back to its actual pin number.

In the meantime though, all of these built-in objects are comparable (i.e. you can use the `in` operator on a list, or list.index) and (as of a few days ago in the nightly builds) hashable (so they can be keys in dictionaries). So given a statemachine instance, there are several easy ways to find out which one it is.

**Comment by GuillaumeLeclerc** [comment]
Thanks for the answer. Are there any efficient ways ? (In viper or inline
assembly). I guess just reading the IRQ register is probably the best way ?

**Comment by GuillaumeLeclerc** [comment]
I just realized this is particularly a problem because it is impossible to have hard interrupts if the handler are closures, so there is really no way in python to identify which state machine is interrupting.

I get 
```
Uncaught exception in IRQ callback handler
MemoryError: 
```
If I try that. Hard interrupts only work when they are a global scope (so no way to have one per state machine without having a copy of the interrupt for every single one.

**Comment by GuillaumeLeclerc** [comment]
It Also seems that the Interrupt Flags are cleared before the handler is called so I'm really not sure how I can possibly know which StateMachine is being triggered (fast).

**Comment by rkompass** [comment]
There is an undocumented functionality in array.array: The 'o' typecode, that allows to inspect the data stored at MPs objects.
```
# Use inspection of RP2 objects to find out the number of the PIO state machine.

from machine import Pin, mem16
from array import array
import rp2
from time import sleep_ms

@rp2.asm_pio(set_init=rp2.PIO.OUT_LOW)
def blink_1hz():
    irq(rel(0))
    set(pins, 1)
    set(x, 31)                  [5]
    label("delay_high")
    nop()                       [29]
    jmp(x_dec, "delay_high")
    set(pins, 0)
    set(x, 31)                  [6]
    label("delay_low")
    nop()                       [29]
    jmp(x_dec, "delay_low")

# Create StateMachine1 with the blink_1hz program, outputting on Pin(9).
sm1 = rp2.StateMachine(0, blink_1hz, freq=2000, set_base=Pin(9))
# Create StateMachine2 with the blink_1hz program, outputting on Pin(10).
sm2 = rp2.StateMachine(5, blink_1hz, freq=2000, set_base=Pin(10))

def inspect(p):
    baddr = bytes(array('O', [p]))
    addr = int.from_bytes(baddr, 'little')
#    print('MP object at: 0x%08x' % addr)
    print('PIO State Machine:', mem16[addr + 10])

# Set the IRQ handler 1 to print the state machine number.
sm1.irq(inspect)
# Set the IRQ handler 2 to print the state machine number.
sm2.irq(inspect)

# Start the StateMachines.
sm1.active(1)
sleep_ms(500)
sm2.active(1)
```
